src/ucp/api/ucp.h

```diff
@@ -3135,6 +3108,185 @@ ucs_status_ptr_t ucp_worker_flush_nb(ucp_worker_h worker, unsigned flags,
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Completion callback for data arrivals.
+ *
+ * This callback routine is invoked whenever the @ref ucp_tag_send_nb
+ * The ucp library will drive a function on the arrival of each fragment
+ * of data from the initiator.
+ *
+ * @param [out} target    Where the data should be placed. NULL if no copying
+ *                        needed.
+ * @param [in}  source    Where the data currently is
+ * @param [in]  bytes     How much data has just arrived
+ * @param [in]  cookie    Data cookie from the ucp_am_recv structure
+ */
+typedef void (*ucp_am_data_function_t)( void *target,
+                                        void *source,
+                                        size_t bytes,
+                                        void *cookie
+                                      );
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Completion callback for transfer from initiator complete.
+ *
+ * The ucp library will drive a function when the transfer from the
+ * initiator is complete.
+ *
+ * @param [in}  arg             Request structure
+ * @param [in}  cookie          Cookie from the ucp_am_recv structure
+ * @param [in]  iovec           Description of the received data
+ * @param [in]  iovec_length    Number of elements in iovec
+ */
+typedef ucs_status_t (*ucp_am_local_function_t)( void *arg,
+                                                 void *cookie,
+                                                 ucp_dt_iov_t *iovec,
+                                                 size_t iovec_length
+                                               );
+/*
+ * Structure for communication between the active message
+ * callback and the ucp library. The ucp library sets up 'iovec_max_length'
+ * to indicate the length of the iovec in this structure; the remaining
+ * fields are set by the callback to indicate what should happen
+ * with the remaining data from the initiator. In the initial implementation,
+ * iovec_max_length is always 1.
+ */
+typedef struct ucp_am_recv {
+    /* Function to be driven when data transfer is complete */
+    ucp_am_local_function_t local_fn;
+    /* Argument to be passed to local_fn */
+    void                   *cookie;
+    /* Function to be driven when each fragment of data transfer is complete
+     * In the initial implementation, all data is transferred in one fragment,
+     * so the data_fn is called once just before the local_fn .
+     * This implementation is to ease porting of applicaitons currently
+     * coded to use IBM PAMI.
+     */
+    ucp_am_data_function_t  data_fn;
+    /* Argument to be passed to data_fn */
+    void                   *data_cookie;
+    /* Sise of iovec allocated by the ucp library */
+    size_t                  iovec_max_length;
+    /* Size of iovec filled in by the callback */
+    size_t                  iovec_length;
+    /* iovec indicating where the data from the initiator is to be placed */
+    ucp_dt_iov_t            iovec[1];
```

Comment:
the struct is not extendable because it does not have field mask and also this array is in the end